File: e-scala/src/main/scala/dev/akif/e/scala/Codec.scala
package dev.akif.e.scala

import dev.akif.e.{AbstractCodec, AbstractDecoder, AbstractEncoder}

trait Decoder[IN] extends AbstractDecoder[IN, E]

trait Encoder[OUT] extends AbstractEncoder[E, OUT]

trait Codec[IN, OUT] extends AbstractCodec[E, IN, OUT]

File: e-kotlin/src/main/kotlin/dev/akif/e/kotlin/Codec.kt
package dev.akif.e.kotlin

import dev.akif.e.AbstractCodec
import dev.akif.e.AbstractDecoder
import dev.akif.e.AbstractEncoder

interface Codec<A> : AbstractCodec<E, A>

interface Decoder<IN> : AbstractDecoder<IN, E>

interface Encoder<OUT> : AbstractEncoder<E, OUT>

fun <IN> Decoder<IN>.decodeMaybe(input: IN): Maybe<E> {
    val result = this.decode(input)
    val e      = result.get()

    return when {
        !result.isSuccess -> Failure(e)
        else              -> Success(e)
    }
}
